### DISCUSSION TITLE
Call a function when the task is scheduled

### DISCUSSION THREAD

**Discussion by a1d4r** [taskiq-python/taskiq#308] [Q&A]
Hi! I'm implementing custom `ScheduleSource` which loads schedules from PostgreSQL. I would like to allow a user to set up schedules for tasks so that the scheduler loads them dynamically. It seems to be not difficult to implement `get_schedules` method which is going to query the database.

I also would like user to see the tasks which are going to be executed in the future (one upcoming task for each schedule). For such case, there is `runs` table in the database which stores records about past and upcoming runs. However, I couldn't find a solution to implement it using taskiq scheduler. It would be great if `ScheduleSource` had a method which is called by the scheduler when the task is scheduled to be executed in the future (when the coroutine with delayed send is created). 

I would like to contribute to the library to extend the scheduler:
- extend [ScheduleSource](https://github.com/taskiq-python/taskiq/blob/develop/taskiq/abc/schedule_source.py#L8) class with a method called `on_schedule`. 
- extend [TaskiqScheduler](https://github.com/taskiq-python/taskiq/blob/develop/taskiq/scheduler/scheduler.py#L16) with a method called `on_schedule`.
- In [delayed_send](https://github.com/taskiq-python/taskiq/blob/develop/taskiq/cli/scheduler/run.py#L106) call `scheduler.on_schedule(source, task)` before going to sleep

Does it make sense?

**Comment by s3rius** [comment]
So, basically it's not possible, because tasks are not planned to be executed before the actual execution. The scheduler grabs all schedules  from all sources and calculates which schedules to run now. 

But, you can easily get all planned tasks, by filtering out all tasks which has `time` value less than the current time. All tasks with cron should not be filtered, because they are reccurent.
